tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,46 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+
+  // if this is the Entry FD, then add the target profile
+  // shader attribute to the FD and carry on with validation
+  // otherwise, with no shader attribute, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then there's no
+  // shader attribute to work with, so just return
+  if (EntryPointName != FD->getIdentifier()->getName() || 
+    (EntryPointName == FD->getIdentifier()->getName() && !FD->isGlobal())) {
```

Comment:
I don't think `!FD->isGlobal()` is quite right, however, I don't think the condition on the call to `DiagnoseEntry()` is sufficient in the first place.  I think we need to check a few more conditions to ensure we are only calling `DiagnoseEntry()` on top-level function definitions (not declarations) that really could be entry points.  

For instance, I think this condition should be sufficient (`SemaDecl.cpp` near end of `Sema::ActOnFunctionDeclarator`):
```cpp
  if (getLangOpts().HLSL && D.isFunctionDefinition() && D.hasName() &&
      NewFD->getDeclContext()->getRedeclContext()->isTranslationUnit()) {
    hlsl::DiagnoseEntry(*this, NewFD);
  }
```

With that in place, `TryAddShaderAttrFromTargetProfile` doesn't need to check `isGlobal()`.